[dynamo][guards] Skip guards on empty nn module hooks

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139038
* __->__ #138942
* #138954
* #139040

This brings some unsoundness in guards. Earlier we were skipping empty nn module hooks dict guard only on inbuilt nn modules, but as seen in https://github.com/pytorch/pytorch/issues/138386, there could be still be significant guard overhead. With this PR, we reduce the guard eval latency from 420 us to 280 us (1.5x reduction).


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

anijain2305 committed:
[dynamo][guards] Skip guards on empty nn module hooks

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Skip guards on empty nn module hooks"


This brings some unsoundness in guards. Earlier we were skipping empty nn module hooks dict guard only on inbuilt nn modules, but as seen in https://github.com/pytorch/pytorch/issues/138386, there could be still be significant guard overhead. With this PR, we reduce the guard eval latency from 420 us to 280 us (1.5x reduction).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Skip guards on empty nn module hooks"


This brings some unsoundness in guards. Earlier we were skipping empty nn module hooks dict guard only on inbuilt nn modules, but as seen in https://github.com/pytorch/pytorch/issues/138386, there could be still be significant guard overhead. With this PR, we reduce the guard eval latency from 420 us to 280 us (1.5x reduction).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -470,6 +470,7 @@ def fn(x):
         self._validate(fn, backend, x, skip_check=True)
 
     @torch._functorch.config.patch(recompute_views=True)
+    @torch._dynamo.config.patch(skip_nnmodule_hook_guards=False)
```
@yf225 I am not sure if you wrote the test with Dynamo identifying that a hook has been applied, and therefore it needs to recompile. You introduced this in https://github.com/pytorch/pytorch/pull/135727

anijain2305 committed:
Update on "[dynamo][guards] Skip guards on empty nn module hooks"


This brings some unsoundness in guards. Earlier we were skipping empty nn module hooks dict guard only on inbuilt nn modules, but as seen in https://github.com/pytorch/pytorch/issues/138386, there could be still be significant guard overhead. With this PR, we reduce the guard eval latency from 420 us to 280 us (1.5x reduction).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -519,6 +519,7 @@ def _factory_fn():
                 mod_no_hook, backend, x, fullgraph=True, compiled_autograd=True
             )
 
+        torch._dynamo.reset()
```
@yf225 From what I understand, this test was written to check that the partitioner is able to save the new intermediate from the backward hook.

However, the way the test is written, currently it relies on Dynamo to see that a new hook has been applied to the module and recompile. With the change in this PR, I am bringing some unsoundness because this gives quite good speedup in general. So, I am wondering if this is a common case where you compile the model and then apply the hook? Or this is just a co-incidence and was not really the intention of this test.

yf225 replied:
I believe it's just a co-incidence of how the test is written and was not the intention of this test (it's more meant to have an entirely new module to register a hook on).

I believe as long as the check `self.assertEqual(additional_saved_tensors, expected_additional_saved_tensors)` can still pass (which relies on the Inductor post_grad pass `collect_fwd_graph_outputs` to collect the saved tensors) we are good here

ezyang commented:
Are you going to allow controlling the unsoundness via config?

anijain2305 commented:
> Are you going to allow controlling the unsoundness via config?

Yes, `skip_nnmodule_hook_guards` controls this.

ezyang commented:
I don't see any reference to it in this PR, is there some nontrivial interaction with earlier PRs in the stack?

anijain2305 commented:
> I don't see any reference to it in this PR, is there some nontrivial interaction with earlier PRs in the stack?

Its not obvious from this PR. The change in the PR returns a EMPTY_NN_MODULE_HOOKS_DICT guard, which relies on the `skip_nnmodule_hook_guards` flag to skip. 
https://github.com/pytorch/pytorch/blob/d2052ea84d8e8d52c1c8c32269851bcd11ad2aed/torch/_dynamo/guards.py#L1681-L1686


Let me also add a test just to make it more clear.

anijain2305 commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -1080,9 +1080,7 @@ def var_getattr(self, tx: "InstructionTranslator", name):
-            if not tx.output.side_effects.has_pending_mutation_of_attr(
-                self, name
-            ) and self.value.__module__.startswith(("torch.nn.", "torch.ao.")):
+            if not tx.output.side_effects.has_pending_mutation_of_attr(self, name):
```
Add test

anijain2305 reviewed (commented):
\

anijain2305 committed:
Update on "[dynamo][guards] Skip guards on empty nn module hooks"


This brings some unsoundness in guards. Earlier we were skipping empty nn module hooks dict guard only on inbuilt nn modules, but as seen in https://github.com/pytorch/pytorch/issues/138386, there could be still be significant guard overhead. With this PR, we reduce the guard eval latency from 420 us to 280 us (1.5x reduction).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Skip guards on empty nn module hooks"


This brings some unsoundness in guards. Earlier we were skipping empty nn module hooks dict guard only on inbuilt nn modules, but as seen in https://github.com/pytorch/pytorch/issues/138386, there could be still be significant guard overhead. With this PR, we reduce the guard eval latency from 420 us to 280 us (1.5x reduction).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] Skip guards on empty nn module hooks"


This brings some unsoundness in guards. Earlier we were skipping empty nn module hooks dict guard only on inbuilt nn modules, but as seen in https://github.com/pytorch/pytorch/issues/138386, there could be still be significant guard overhead. With this PR, we reduce the guard eval latency from 420 us to 280 us (1.5x reduction).


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge -f "unrelated CI failure"

ezyang commented:
This seems to result in a modest compile time improvement in HF inference, probably because processing all the extra guards is also expensive.

<img width="387" alt="image" src="https://github.com/user-attachments/assets/2ec595c3-d336-44a2-a034-3b782fa9637d">

Samples: https://fburl.com/scuba/torch_open_source_signpost/5snsgy3x
